src/util/PersistentStorage.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "daisy_core.h"
+#include "per/qspi.h"
+
+namespace daisy
+{
+/** @brief Non Volatile storage class for persistent settings on an external flash device.
+ *  @author shensley
+ * 
+ *  Storage occupied by the struct will be one word larger than 
+ *  the SettingStruct used. The extra word is used to store the
+ *  state of the data, and whether it's been overwritten or not.
+ * 
+ *  \todo - Make Save() non-blocking
+ *  \todo - Add wear leveling
+ * 
+ **/
+template <typename SettingStruct>
+class PersistentStorage
+{
+  public:
+    enum class State
+    {
+        UNKNOWN = 0,
+        FACTORY = 1,
+        USER    = 2,
+    };
+
+    PersistentStorage() : settings_(nullptr) {}
+
+    /** Initialize Storage class
+     *
+     *  The values in this class will be stored as the default
+     *  for restoration to 'factory' settings.
+     *
+     *  \param qspi handle to the hardware qspi peripheral.
+     *  \param settings should be a setting structure containing the default values.
+     *      this will be updated to contain the stored data.
+     *  \param start_address address for location on the QSPI chip (offset to base address of device).
+     *      This defaults to the first address on the chip, and should be a multiple of 256
+     **/
+    void Init(QSPIHandle *   qspi,
+              SettingStruct &settings,
+              uint32_t       start_address = 0x90000000)
+    {
+        qspi_             = qspi;
+        state_            = State::UNKNOWN;
+        default_settings_ = settings; // copy data to fixed struct.
+        settings_ = &settings; // copy reference to pointer for modification.
```

Comment:
```suggestion
    void Init(QSPIHandle&          qspi,
              const SettingStruct& defaults,
              uint32_t             start_address = kMemBaseAddr)
    : qspi_(qspi)
    , start_address_(start_address)
    , default_settings_(defaults)
    , settings_(defaults)
    , state_(State::UNKNOWN)
    {
```

Reply:
woah, I didn't know you could initializer lists in non-constructor functions. that's interesting.

Reply:
Oopsi. I was wrong, I misinterpreted this as the constructor. Doh. Sorry ...

Reply:
Haha, no problem. I feel like I still learn about some new language feature every other week. So I wouldn't have been surprised if you could.